### PR TITLE
[Feature] Turns a gitfs remote into a true "formula", merges specified branch into current saltenv

### PR DESCRIPTION
**Pull request by ekristen**
Thank you to @terminalmage for taking the time to give me basically everything I needed to implement this in a comment on the original issue. Very much appreciated.

**Note:** (added after @terminalmage initial review comment) -- we originally discussed naming the variable `global` but that seems to be reserved and causes python syntax errors when being referenced on `self` (ie `return self.global`). I used `formula` for the initial PR, but I agree it's not the best name as this feature has many more uses. 

### What does this PR do?
It enables a gitfs remote to be a true formula, or a global one, when set, all states from the referenced branch are merged into whichever saltenv is actually being referenced by the minion in which the states are being resolved for.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/40527

### How to use and test this feature.
Add the following to your salt-minion (for a masterless test) config.

```yaml
environment: testing
fileserver_backend:
  - git
gitfs_remotes:
  - https://github.com/ekristen/salt-states-testing.git
  - https://github.com/ekristen/salt-formula-testing.git
```

If you inspect `https://github.com/ekristen/salt-states-testing.git` you'll find a testing branch and if you inspect `https://github.com/ekristen/salt-formula-testing.git` you will NOT find a testing branch. Given this setup, if you run the following command `salt-call --local state.sls testing` it will succeed (as it is from the first repo), however if you run `salt-call --local state.sls formula` it will fail as there is no `testing` branch on the second repository.

Modify the config so it looks like this.

```yaml
environment: testing
fileserver_backend:
  - git
gitfs_remotes:
  - https://github.com/ekristen/salt-states-testing.git
  - https://github.com/ekristen/salt-formula-testing.git:
    - formula: master
```

Then rerun the commands, first `salt-call --local state.sls testing` it will still resolve and run, and now `salt-call --local state.sls formula` this time it will resolve and run appropriately. 

### Tests written?
No - I looked at the gitfs tests and there did not seem to be an appropriate way or place to test this new feature within the current test suite for gitfs.

### CODE REVIEW

File: salt/fileserver/gitfs.py
Comment (terminalmage) [inline salt/fileserver/gitfs.py]: I thought we discussed naming this as ``global``, not ``formula``. This feature is useful for more than just formulas, so giving it such a specific name seems wrong.

**Review by terminalmage — CHANGES_REQUESTED** [review]

File: salt/fileserver/gitfs.py
Comment (ekristen) [inline salt/fileserver/gitfs.py]: @terminalmage my bad! I forgot to mention when I typed up the original PR, I tried global, but it kept throwing a syntax error, seems to be a reserved keyword.

Comment (ekristen) [inline salt/fileserver/gitfs.py]: Perhaps a synonym? `world`, `wordwide`, or `universal` (I think I like universal the best?)

Comment (terminalmage) [inline salt/fileserver/gitfs.py]: Oh, interesting. I've never tried using ``global`` as a class attribute, so I've never run into this syntax error before. I'll try to brainstorm some ideas as well.

Comment (terminalmage) [inline salt/fileserver/gitfs.py]: What about ``all_saltenvs``?

Comment (ekristen) [inline salt/fileserver/gitfs.py]: That works for me, the value would still be the branch name I assume?

Comment (terminalmage) [inline salt/fileserver/gitfs.py]: Yeah. If you're comfortable writing RST, this should be documented in the gitfs documentation page at doc/topics/tutorials/gitfs.rst. If not, I can take care of documentation.

**Comment by terminalmage** [comment]
Don't worry about tests, we have limited test coverage for gitfs but I've been working on a framework which will make it easier to do integration tests on git_pillar and gitfs (https://github.com/saltstack/salt/pull/40777). Writing new integration tests for gitfs is something I have plans for soon anyway, and I can make sure something like this is covered.

File: salt/fileserver/gitfs.py
Comment (ekristen) [inline salt/fileserver/gitfs.py]: I'm fine with writing RST. I can take a stab at adding some initial docs in.

Comment (terminalmage) [inline salt/fileserver/gitfs.py]: Cool. I can help polish them if they need to be tweaked in any way.

Thanks for taking this on, I'm really excited about this feature!

Comment (ekristen) [inline salt/fileserver/gitfs.py]: Glad to help, this will be very useful for me.

**Comment by ekristen** [comment]
@terminalmage I did notice the original issue was slated for `Oxygen` but I assume that was based on if you had to do all the work, with an open PR now, might it be possible to get it merged for an upcoming release?

For now I left all documentation notes to reference Oxygen.

**Comment by terminalmage** [comment]
We've passed the feature freeze for Nitrogen, so develop is the correct branch and Oxygen is the correct release.

**Comment by ekristen** [comment]
Is there a timeline? https://docs.saltstack.com/en/latest/topics/releases/version_numbers.html has TBD for both Nitrogen and Oxygen. Thanks!

**Comment by terminalmage** [comment]
Nitrogen is expected late this month, with Oxygen in the fall (tentatively).

**Comment by cachedout** [comment]
@terminalmage Can you re-review this and approve if you think it's ready?

**Comment by terminalmage** [comment]
@ekristen I opened https://github.com/ekristen/salt/pull/1 against your fork which tweaks and builds upon the docs you added, and also modifies the gitfs code to add debug logging when a per-saltenv configuration is being overridden by ``all_saltenvs``. This is something I had not considered initially until I started testing the functionality of your PR.

Can you please merge that PR and then test and confirm that it still works to your liking? If so, and assuming tests pass, I think this is good to merge.

Thanks once again for your contribution.

**Comment by ekristen** [comment]
@terminalmage I have merged your PR, I will test shortly and confirm all is working appropriately still.

**Comment by ekristen** [comment]
@terminalmage everything works, the only comment I have is that the debug messages off a bit?

They seem to repeat multiple times, and always state `None` was mapped to `saltenv`. As you are way more familiar with the system I'll defer to you as to whether or not this is correct, just wanted to mention it.

```
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'base' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'testing' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
[DEBUG   ] The per-saltenv configuration has mapped the 'None' branch/tag to saltenv 'testing' for gitfs remote 'https://github.com/ekristen/salt-formula-testing.git', but this remote has all_saltenvs set to 'master'. The per-saltenv mapping will be ignored in favor of 'master'.
```

**Comment by terminalmage** [comment]
That definitely is not correct. Change this line:

```python
if all_saltenvs_ref != per_saltenv_ref:
```

to

```python
if per_saltenv_ref and all_saltenvs_ref != per_saltenv_ref:
```

I didn't test it without a per-saltenv config, duh.  Thanks for catching this.

**Comment by cachedout** [comment]
@terminalmage Looks like the requested change has been made. Requesting re-review.

**Review by terminalmage — APPROVED** [review]
LGTM. Thanks, @ekristen!

**Comment by ekristen** [comment]
Thanks @terminalmage

**Pull request merged**
